File: envoy-xds/xds/java-xds-impl/src/main/java/indi/yuluo/envoy/xds/core/CallBack.java
package indi.yuluo.envoy.xds.core;

import io.envoyproxy.controlplane.server.DiscoveryServerCallbacks;
import io.envoyproxy.controlplane.server.exception.RequestException;
import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryRequest;
import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryResponse;
import io.envoyproxy.envoy.service.discovery.v3.DiscoveryRequest;
import io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse;

/**
 * @author yuluo
 * @author 1481556636@qq.com
 */

public class CallBack implements DiscoveryServerCallbacks {

	@Override
	public void onStreamClose(long streamId, String typeUrl) {

		System.out.println("onStreamClose exec ... id " + streamId + " typeUrl " + typeUrl);
	}

	@Override
	public void onStreamCloseWithError(long streamId, String typeUrl, Throwable error) {

		System.out.println("onStreamCloseWithError exec ... id " + streamId + " typeUrl " + typeUrl);
	}

	@Override
	public void onStreamOpen(long streamId, String typeUrl) throws RequestException {

		System.out.println("onStreamOpen exec ... id " + streamId + " typeUrl " + typeUrl);
	}

	@Override
	public void onV3StreamRequest(long l, DiscoveryRequest discoveryRequest) throws RequestException {

		System.out.println("onV3StreamRequest exec ... id " + l);
	}

	@Override
	public void onV3StreamDeltaRequest(long l, DeltaDiscoveryRequest deltaDiscoveryRequest) throws RequestException {

		System.out.println("onV3StreamDeltaRequest exec ... id " + l);
	}

	@Override
	public void onV3StreamResponse(long streamId, DiscoveryRequest request, DiscoveryResponse response) {

		System.out.println("onV3StreamResponse exec ... id " + streamId);
	}

	@Override
	public void onV3StreamDeltaResponse(long streamId, DeltaDiscoveryRequest request, DeltaDiscoveryResponse response) {

		System.out.println("onV3StreamDeltaResponse exec ... id " + streamId);
	}
}

File: envoy-xds/xds/golang-xds-impl/middleware/call_back.go
package middleware

import (
	"context"
	"golang-xds-impl/pkg/log"

	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
)

// callback 回调日志信息

type CallBacksMiddleWares struct {
	XLog log.XLogger
}

func (cbm CallBacksMiddleWares) OnFetchRequest(ctx context.Context, request *discovery.DiscoveryRequest) error {

	if cbm.XLog.Debug {
		cbm.XLog.Infof("OnFetchRequest exec ....")
	}

	return nil
}

func (cbm CallBacksMiddleWares) OnFetchResponse(request *discovery.DiscoveryRequest, response *discovery.DiscoveryResponse) {
	if cbm.XLog.Debug {
		cbm.XLog.Infof("OnFetchResponse exec ....")
	}
}

func (cbm CallBacksMiddleWares) OnStreamOpen(ctx context.Context, id int64, types string) error {
	if cbm.XLog.Debug {
		cbm.XLog.Infof("OnStreamOpen exec .... id is %d, type is %v", id, types)
	}

	return nil
}

func (cbm CallBacksMiddleWares) OnStreamClosed(id int64, node *corev3.Node) {

	if cbm.XLog.Debug {
		cbm.XLog.Infof("OnStreamClosed exec .... stream id is %d, node is %s", id, node.Id)
	}
}

func (cbm CallBacksMiddleWares) OnStreamRequest(id int64, request *discovery.DiscoveryRequest) error {

	if cbm.XLog.Debug {
		cbm.XLog.Infof("OnStreamRequest exec ....")
	}

	return nil
}

func (cbm CallBacksMiddleWares) OnStreamResponse(ctx context.Context, id int64, request *discovery.DiscoveryRequest, response *discovery.DiscoveryResponse) {

	if cbm.XLog.Debug {

		cbm.XLog.Infof("OnStreamResponse exec .... id is %d.", id)
	}
}

func (cbm CallBacksMiddleWares) OnDeltaStreamOpen(ctx context.Context, id int64, types string) error {

	if cbm.XLog.Debug {

		cbm.XLog.Infof("OnDeltaStreamOpen exec .... id is %d, type is %v", id, types)
	}

	return nil
}

func (cbm CallBacksMiddleWares) OnDeltaStreamClosed(id int64, node *corev3.Node) {

	if cbm.XLog.Debug {

		cbm.XLog.Infof("OnDeltaStreamClosed exec ....")
	}
}

func (cbm CallBacksMiddleWares) OnStreamDeltaRequest(id int64, request *discovery.DeltaDiscoveryRequest) error {

	if cbm.XLog.Debug {

		cbm.XLog.Infof("OnStreamDeltaRequest exec ....")
	}

	return nil
}

func (cbm CallBacksMiddleWares) OnStreamDeltaResponse(id int64, request *discovery.DeltaDiscoveryRequest, response *discovery.DeltaDiscoveryResponse) {

	if cbm.XLog.Debug {

		cbm.XLog.Infof("OnStreamDeltaResponse exec ....")
	}
}
